### ISSUE TITLE
Performance regression when using rector 1.05 with phpstan 1.11.0

### ISSUE THREAD

**Issue by dorrogeray** [rectorphp/rector#8638]
# Bug Report

Looks like that there is some kind of performance issue, perhaps a memory leak. It manifests itself when I use a large number of processes, like 30. There also seems to be increased memory usage & the initial start of the workers seems delayed (progress bar is stuck at 0 for long time).

Sometimes, I even get this: 
![image](https://github.com/rectorphp/rector/assets/4999006/4b19269a-f0e6-4dfa-b555-b42d042a3002)

| Subject        | Details                                                         |
| :------------- | :---------------------------------------------------------------|
| Rector version | v1.0.5 |

```sh
www-data@4486ce23c81e:~/html$ vendor/bin/rector process
  3170/12825 [▓▓▓▓▓▓░░░░░░░░░░░░░░░░░░░░░░]  24%[14-May-2024 15:47:27 EUROPE/LONDON] PHP Fatal error:  Uncaught Error: Call to a member function end() on null in /var/www/html/vendor/rector/rector/vendor/symplify/easy-parallel/src/ValueObject/ParallelProcess.php:120
Stack trace:
#0 /var/www/html/vendor/rector/rector/vendor/symplify/easy-parallel/src/ValueObject/ProcessPool.php(49): RectorPrefix202405\Symplify\EasyParallel\ValueObject\ParallelProcess->quit()
#1 /var/www/html/vendor/rector/rector/vendor/symplify/easy-parallel/src/ValueObject/ProcessPool.php(59): RectorPrefix202405\Symplify\EasyParallel\ValueObject\ProcessPool->quitProcess()
#2 /var/www/html/vendor/rector/rector/src/Parallel/Application/ParallelFileProcessor.php(108): RectorPrefix202405\Symplify\EasyParallel\ValueObject\ProcessPool->quitAll()
#3 /var/www/html/vendor/rector/rector/vendor/symplify/easy-parallel/src/ValueObject/ParallelProcess.php(108): Rector\Parallel\Application\ParallelFileProcessor->Rector\Parallel\Application\{closure}()
#4 /var/www/html/vendor/rector/rector/vendor/react/event-loop/src/Timer/Timers.php(86): RectorPrefix202405\Symplify\EasyParallel\ValueObject\ParallelProcess->RectorPrefix202405\Symplify\EasyParallel\ValueObject\{closure}()
#5 /var/www/html/vendor/rector/rector/vendor/react/event-loop/src/StreamSelectLoop.php(150): RectorPrefix202405\React\EventLoop\Timer\Timers->tick()
#6 /var/www/html/vendor/rector/rector/src/Parallel/Application/ParallelFileProcessor.php(178): RectorPrefix202405\React\EventLoop\StreamSelectLoop->run()
#7 /var/www/html/vendor/rector/rector/src/Application/ApplicationFileProcessor.php(237): Rector\Parallel\Application\ParallelFileProcessor->process()
#8 /var/www/html/vendor/rector/rector/src/Application/ApplicationFileProcessor.php(136): Rector\Application\ApplicationFileProcessor->runParallel()
#9 /var/www/html/vendor/rector/rector/src/Console/Command/ProcessCommand.php(117): Rector\Application\ApplicationFileProcessor->run()
#10 /var/www/html/vendor/rector/rector/vendor/symfony/console/Command/Command.php(327): Rector\Console\Command\ProcessCommand->execute()
#11 /var/www/html/vendor/rector/rector/vendor/symfony/console/Application.php(960): RectorPrefix202405\Symfony\Component\Console\Command\Command->run()
#12 /var/www/html/vendor/rector/rector/vendor/symfony/console/Application.php(333): RectorPrefix202405\Symfony\Component\Console\Application->doRunCommand()
#13 /var/www/html/vendor/rector/rector/src/Console/ConsoleApplication.php(53): RectorPrefix202405\Symfony\Component\Console\Application->doRun()
#14 /var/www/html/vendor/rector/rector/vendor/symfony/console/Application.php(216): Rector\Console\ConsoleApplication->doRun()
#15 /var/www/html/vendor/rector/rector/bin/rector.php(130): RectorPrefix202405\Symfony\Component\Console\Application->run()
#16 /var/www/html/vendor/rector/rector/bin/rector(5): require_once('...')
#17 /var/www/html/vendor/bin/rector(119): include('...')
#18 {main}
  thrown in /var/www/html/vendor/rector/rector/vendor/symplify/easy-parallel/src/ValueObject/ParallelProcess.php on line 120
Fatal error: Uncaught Error: Call to a member function end() on null in /var/www/html/vendor/rector/rector/vendor/symplify/easy-parallel/src/ValueObject/ParallelProcess.php:120
Stack trace:
#0 /var/www/html/vendor/rector/rector/vendor/symplify/easy-parallel/src/ValueObject/ProcessPool.php(49): RectorPrefix202405\Symplify\EasyParallel\ValueObject\ParallelProcess->quit()
#1 /var/www/html/vendor/rector/rector/vendor/symplify/easy-parallel/src/ValueObject/ProcessPool.php(59): RectorPrefix202405\Symplify\EasyParallel\ValueObject\ProcessPool->quitProcess()
#2 /var/www/html/vendor/rector/rector/src/Parallel/Application/ParallelFileProcessor.php(108): RectorPrefix202405\Symplify\EasyParallel\ValueObject\ProcessPool->quitAll()
#3 /var/www/html/vendor/rector/rector/vendor/symplify/easy-parallel/src/ValueObject/ParallelProcess.php(108): Rector\Parallel\Application\ParallelFileProcessor->Rector\Parallel\Application\{closure}()
#4 /var/www/html/vendor/rector/rector/vendor/react/event-loop/src/Timer/Timers.php(86): RectorPrefix202405\Symplify\EasyParallel\ValueObject\ParallelProcess->RectorPrefix202405\Symplify\EasyParallel\ValueObject\{closure}()
#5 /var/www/html/vendor/rector/rector/vendor/react/event-loop/src/StreamSelectLoop.php(150): RectorPrefix202405\React\EventLoop\Timer\Timers->tick()
#6 /var/www/html/vendor/rector/rector/src/Parallel/Application/ParallelFileProcessor.php(178): RectorPrefix202405\React\EventLoop\StreamSelectLoop->run()
#7 /var/www/html/vendor/rector/rector/src/Application/ApplicationFileProcessor.php(237): Rector\Parallel\Application\ParallelFileProcessor->process()
#8 /var/www/html/vendor/rector/rector/src/Application/ApplicationFileProcessor.php(136): Rector\Application\ApplicationFileProcessor->runParallel()
#9 /var/www/html/vendor/rector/rector/src/Console/Command/ProcessCommand.php(117): Rector\Application\ApplicationFileProcessor->run()
#10 /var/www/html/vendor/rector/rector/vendor/symfony/console/Command/Command.php(327): Rector\Console\Command\ProcessCommand->execute()
#11 /var/www/html/vendor/rector/rector/vendor/symfony/console/Application.php(960): RectorPrefix202405\Symfony\Component\Console\Command\Command->run()
#12 /var/www/html/vendor/rector/rector/vendor/symfony/console/Application.php(333): RectorPrefix202405\Symfony\Component\Console\Application->doRunCommand()
#13 /var/www/html/vendor/rector/rector/src/Console/ConsoleApplication.php(53): RectorPrefix202405\Symfony\Component\Console\Application->doRun()
#14 /var/www/html/vendor/rector/rector/vendor/symfony/console/Application.php(216): Rector\Console\ConsoleApplication->doRun()
#15 /var/www/html/vendor/rector/rector/bin/rector.php(130): RectorPrefix202405\Symfony\Component\Console\Application->run()
#16 /var/www/html/vendor/rector/rector/bin/rector(5): require_once('...')
#17 /var/www/html/vendor/bin/rector(119): include('...')
#18 {main}
  thrown in /var/www/html/vendor/rector/rector/vendor/symplify/easy-parallel/src/ValueObject/ParallelProcess.php on line 120
```

## Minimal PHP Code Causing Issue

Not related to particular code

## Expected Behaviour

Rector works fine when used in parallel mode

**Comment by samsonasik** [comment]
On big project, that may similar with https://github.com/rectorphp/rector/issues/8637

I will look into it, closing as possible known issue on very big project and may stale as may require some architectural change, ensuring call `->reflectAllClasses()` once instead of call in total workers.

You may can configure parallel config for now with add more timeout value, and less maxNumberOfProcess for now.

**Comment by samsonasik** [comment]
I created PR:

- https://github.com/rectorphp/rector-src/pull/5877

for it.

**Comment by samsonasik** [comment]
Closing for now to avoid distractions and conflict, and better peace in my mind.

I am aware of it, and will revisit later with hopefully better solution.

**Comment by samsonasik** [comment]
I created new PR:

- https://github.com/rectorphp/rector-src/pull/5878

for it.

**Comment by dorrogeray** [comment]
@samsonasik I have installed `dev-main#3c1e41f0062bd1f662a545122174b1fcdf761499` and can confirm its much better now. :+1:

Thank you for the quick resolution! :heart:

**Comment by dorrogeray** [comment]
@samsonasik one more update. I went through our CI history, it turns out that before the update, without any cache, it took about 5.5 minutes to run rector for the project. Right now its 17 minutes when using `dev-main#3c1e41f0062bd1f662a545122174b1fcdf761499`, so while its definitely better than `1.0.5`, its roughly ~3x slower than `1.0.4`.

**Comment by samsonasik** [comment]
Thank you @dorrogeray for the check, I tried looking if the cache can be hit set the service that used it at `FamilyRelationsAnalyzer`:

- https://github.com/rectorphp/rector-src/pull/5882

but on debugging, I got init `reflectAllClasses()` can possibly called more than once. 

I got other priority to work right now, I will look more on other solution/improvement when I have a chance :)

**Comment by X-Coder264** [comment]
Is there any chance of tagging 1.0.6 soon with this fix?

**Comment by staabm** [comment]
Can someone share a reproducer for this perf problem?

**Comment by samsonasik** [comment]
@staabm it mostly already handled at PRs:

- https://github.com/rectorphp/rector-src/pull/5878
- https://github.com/rectorphp/rector-src/pull/5879

For note: the init class name collection cache needs on `ApplicationFileProcessor` level as for now, not in service, as they consumed by `worker` to avoid concurrent request which possibly make multiple cache inits.

I tested on rector-src code base with `2180 files`, it already increase performance significantly:

**Before:** 58.689 seconds
**After:** 37.014 seconds

It may only noticeably on very big project.

**Comment by staabm** [comment]
how fast was the rector job on your machine with 1.0.4?

**Comment by samsonasik** [comment]
I can see `34.333 seconds` vs `32.217 seconds` on latest try on main branch, so only 2 seconds different, so it seems fine on rector-src code base itself.

**Comment by samsonasik** [comment]
@dorrogeray @X-Coder264 Rector 1.1.0 is tagged https://github.com/rectorphp/rector/releases/tag/1.1.0

**Comment by TomasVotruba** [comment]
Reopening to have this in attention, as there are related reports poping.

@dorrogeray Hi, could you share your numbers now? :)

**Comment by samsonasik** [comment]
Here the summary of child detection usage on v1.1.0

**1. On ApplicationFileProcessor Level:**

When there is/are still file(s) to be processed, it do:

- clear class collection names (because new class may be introduced)
- trigger class collection names via `reflectAllClasses()`, only **once**, and when file are more than 1.

Only that, which only take 2 or 3 seconds on rector-src codebase with `2228` files, this, only collect on configured `paths()` or path provided in CLI.

**2. On Worker Level**

This happen specifically on `FamilyRelationsAnalyzer` service on `getChildrenOfClassReflection` method, only when the method called:

- on final class -> no child, no need to load classes
- on non-final classes -> call `loadClasses()`, that verify
    - it cached already, no need to loop again
    - loop classes and call `$this->reflectionProvider->getClass(...)`
- get the child classes of target `ClassReflection`

Possible Improvement
----------------------

- Make property on `FamilyRelationsAnalyzer` to cache child ClassReflections per ClassReflection name, so on multiple rules, it doesn't need to loop class reflections from ReflectionProvider.

For note
--------

The currrent optimization should already provide a lot of performance improvement, but  @dorrogeray would be great if we can see see the number or if we can see the reproducible repo you can share, that would be better to more valid check ;)

**Comment by dorrogeray** [comment]
Hi, thanks for the change! Unfortunately I am on a holiday right now and will not be able to test this until early July, but looking forward to doing so when I come back.. 🙂

**Comment by TomasVotruba** [comment]
## Measuments by version

I've done some debugging on quite good and medium sized project. Removed `final` and all return types to trigger the effect.

Here are my data:

```bash
1.0.4

time vendor/bin/rector --dry-run --clear-cache 439.06s user 11.31s system 1209% cpu 37.244 total

1.0.5

time vendor/bin/rector --dry-run --clear-cache  639.78s user 12.78s system 1267% cpu 51.474 total

1.1.0

time vendor/bin/rector --dry-run --clear-cache  601.26s user 12.54s system 1162% cpu 52.801 total

1.1.0 - without get child classes reflection

vendor/bin/rector --dry-run --clear-cache  468.51s user 10.18s system 1083% cpu 44.167 total
```

## Root Cause

This PR introduced child reflection collecting in 1.0.5 https://github.com/rectorphp/rector-src/pull/5735, to fix https://github.com/rectorphp/rector/issues/8557.

While it makes the bug fixed, it introduced quite a huge complexity and performance hit of 15-20 %. On some projects even more, as it seems the slow down is exponential, as report 8x time slower.

<br>

## Trade off

We also made a move to avoid child classes analysis, as slow and not reliable as shown in https://github.com/rectorphp/rector/issues/8439 by @staabm 

Saying that, I think we'll do the trade of here to:

*  mark https://github.com/rectorphp/rector/issues/8557 as wontfix  (untill we find better way without having all classes at once :)), 
* and revert the child classes collection. 
 
To get back the 15-20% performance hit reported in https://github.com/rectorphp/rector/issues/8637, https://github.com/rectorphp/rector/issues/8638, https://github.com/rectorphp/rector/issues/8679

**Comment by TomasVotruba** [comment]
Thanks @samsonasik for quick work :+1: 

Thanks everyone for patience to all :pray:  I'm now travelling and connection is sometime challange to get here. I know I should have check this more deeply sooner.

Resolved in https://github.com/rectorphp/rector-src/pull/5995
